smarts/core/sensors.py

```diff
@@ -1209,3 +1239,91 @@ def closest_point_on_lane(position, lane_id):
+        self._vehicle = vehicle
+        self._road_map = road_map
+        self._lookahead = lookahead
+
+    @staticmethod
+    def _is_signal_type(feature: RoadMap.Feature) -> bool:
+        # XXX:  eventually if we add other types of dynamic features, we'll need to update this.
+        return (
+            feature.type == RoadMap.FeatureType.FIXED_LOC_SIGNAL or feature.is_dynamic
+        )
+
+    def __call__(
+        self,
+        lane: Optional[RoadMap.Lane],
+        lane_pos: RefLinePoint,
+        state,  # VehicleState
+        plan: Plan,
+        provider_state,  # ProviderState
+    ) -> List[SignalObservation]:
+        result = []
+        if not lane:
+            return result
+        upcoming_signals = []
+        half_len = 0.5 * state.dimensions.length
+        # make sure the signal is not behind me...
+        my_bb = lane_pos.s - half_len
+        for feat in lane.features:
+            if not self._is_signal_type(feat):
+                continue
+            for pt in feat.geometry:
+                if lane.offset_along_lane(pt) >= my_bb:
+                    upcoming_signals.append(feat)
+                    break
+        lookahead = self._lookahead - lane.length + lane_pos.s
+        self._find_signals_ahead(lane, lookahead, plan.route, upcoming_signals)
+
+        for signal in upcoming_signals:
+            for actor_state in provider_state.actors:
+                if actor_state.actor_id == signal.feature_id:
+                    signal_state = actor_state
+                    break
+            else:
+                self._logger.warning(
+                    "could not find signal state corresponding with feature_id={signal.feature_id}"
+                )
+                continue
+            assert isinstance(signal_state, SignalState)
+            controlled_lanes = None
+            if signal_state.controlled_lanes:
+                controlled_lanes = [cl.lane_id for cl in signal_state.controlled_lanes]
+            result.append(
+                SignalObservation(
+                    state=signal_state.state,
+                    stop_point=signal_state.stopping_pos,
+                    controlled_lanes=controlled_lanes,
+                    last_changed=signal_state.last_changed,
+                )
+            )
+
+        return result
+
+    def _find_signals_ahead(
+        self,
+        lane: RoadMap.Lane,
+        lookahead: float,
+        route: Optional[RoadMap.Route],
+        upcoming_signals: List[RoadMap.Feature],
+    ):
+        if lookahead <= 0:
+            return
+        for ogl in lane.outgoing_lanes:
+            if route and ogl.road not in route.roads:
+                continue
+            upcoming_signals += [
+                feat for feat in ogl.features if self._is_signal_type(feat)
+            ]
+            lookahead -= lane.length
+            self._find_signals_ahead(ogl, lookahead, route, upcoming_signals)
```

Comment:
I kind of wish we had a convention for marking recursion since this is important for road map scaling.

Reply:
Yeah, that would have been good.  What would you suggest?  I can at least start using that here.